File: src/main/java/com/likelionknu/applyserver/application/service/ApplicationQueryService.java
package com.likelionknu.applyserver.application.service;

import com.likelionknu.applyserver.application.data.dto.response.ApplicationDetailResponse;
import com.likelionknu.applyserver.application.data.dto.response.ApplicationSummaryResponse;
import com.likelionknu.applyserver.application.data.entity.Application;
import com.likelionknu.applyserver.application.data.entity.RecruitAnswer;
import com.likelionknu.applyserver.application.data.exception.ApplicationNotFoundException;
import com.likelionknu.applyserver.application.data.exception.InvalidApplicationAccessException;
import com.likelionknu.applyserver.application.data.exception.UserNotFoundException;
import com.likelionknu.applyserver.application.data.repository.ApplicationRepository;
import com.likelionknu.applyserver.application.data.repository.RecruitAnswerRepository;
import com.likelionknu.applyserver.auth.data.entity.User;
import com.likelionknu.applyserver.auth.data.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ApplicationQueryService {

    private final ApplicationRepository applicationRepository;
    private final UserRepository userRepository;
    private final RecruitAnswerRepository recruitAnswerRepository;

    @Transactional(readOnly = true)
    public List<ApplicationSummaryResponse> getMyApplications(String email) {
        User user = userRepository.findOptionalByEmail(email)
                .orElseThrow(() -> new UserNotFoundException());

        List<Application> applications = applicationRepository.findAllWithRecruitByUserId(user.getId());

        return applications.stream()
                .map(a -> new ApplicationSummaryResponse(
                        a.getId(),
                        a.getRecruit().getTitle(),
                        a.getStatus().name(),
                        a.getRecruit().getStartAt(),
                        a.getRecruit().getEndAt()
                ))
                .toList();
    }

    @Transactional(readOnly = true)
    public ApplicationDetailResponse getApplicationDetail(String email, Long applicationId) {
        User user = userRepository.findOptionalByEmail(email)
                .orElseThrow(() -> new UserNotFoundException());

        Application application = applicationRepository.findById(applicationId)
                .orElseThrow(() -> new ApplicationNotFoundException());

        if (!application.getUser().getId().equals(user.getId())) {
            throw new InvalidApplicationAccessException();
        }

        List<RecruitAnswer> answers =
                recruitAnswerRepository.findByApplicationId(applicationId);

        return new ApplicationDetailResponse(
                application.getId(),
                application.getRecruit().getId(),
                application.getRecruit().getTitle(),
                application.getStatus().name(),
                application.getRecruit().getStartAt(),
                application.getRecruit().getEndAt(),
                application.getSubmittedAt(),
                answers.stream()
                        .map(a -> new ApplicationDetailResponse.ApplicationAnswerResponse(
                                a.getContent().getId(),
                                a.getContent().getQuestion(),
                                a.getAnswer()
                        ))
                        .toList()
        );
    }

}

File: src/main/kotlin/com/likelionknu/applyserver/application/service/ApplicationQueryService.kt
package com.likelionknu.applyserver.application.service

import com.likelionknu.applyserver.application.data.dto.response.ApplicationDetailResponse
import com.likelionknu.applyserver.application.data.dto.response.ApplicationSummaryResponse
import com.likelionknu.applyserver.application.data.exception.InvalidApplicationAccessException
import com.likelionknu.applyserver.application.data.exception.UserNotFoundException
import com.likelionknu.applyserver.application.data.exception.ApplicationNotFoundException
import com.likelionknu.applyserver.application.data.repository.ApplicationRepository
import com.likelionknu.applyserver.application.data.repository.RecruitAnswerRepository
import com.likelionknu.applyserver.auth.data.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ApplicationQueryService(
    private val applicationRepository: ApplicationRepository,
    private val userRepository: UserRepository,
    private val recruitAnswerRepository: RecruitAnswerRepository
) {

    @Transactional(readOnly = true)
    fun getMyApplications(email: String): List<ApplicationSummaryResponse> {
        val user = userRepository.findOptionalByEmail(email)
            .orElseThrow { UserNotFoundException() }

        val applications = applicationRepository.findAllWithRecruitByUserId(user.id!!)

        return applications.map { a ->
            ApplicationSummaryResponse(
                applicationId = a.id!!,
                recruitTitle = a.recruit.title,
                status = a.status!!.name,
                startAt = a.recruit.startAt,
                endAt = a.recruit.endAt
            )
        }
    }

    @Transactional(readOnly = true)
    fun getApplicationDetail(email: String, applicationId: Long): ApplicationDetailResponse {
        val user = userRepository.findOptionalByEmail(email)
            .orElseThrow { UserNotFoundException() }

        val application = applicationRepository.findById(applicationId)
            .orElseThrow { ApplicationNotFoundException() }

        if (application.user.id != user.id) {
            throw InvalidApplicationAccessException()
        }

        val answers = recruitAnswerRepository.findByApplicationId(applicationId)

        return ApplicationDetailResponse(
            applicationId = application.id!!,
            recruitId = application.recruit.id!!,
            recruitTitle = application.recruit.title,
            status = application.status!!.name,
            startAt = application.recruit.startAt,
            endAt = application.recruit.endAt,
            submittedAt = application.submittedAt,
            answers = answers.map { a ->
                ApplicationDetailResponse.ApplicationAnswerResponse(
                    questionId = a.content.id!!,
                    question = a.content.question,
                    answer = a.answer
                )
            }
        )
    }
}